jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java

```diff
@@ -1199,16 +1217,22 @@ public void set(HttpField field)
         {
             if (_current < 0)
                 throw new IllegalStateException();
-            _fields[_current] = field;
+            if (field == null)
+                remove();
+            else
+                _fields[_current] = field;
         }
 
         @Override
         public void add(HttpField field)
         {
-            _fields = Arrays.copyOf(_fields, _fields.length + 1);
-            System.arraycopy(_fields, _cursor, _fields, _cursor + 1, _size++);
-            _fields[_cursor++] = field;
-            _current = -1;
+            if (field != null)
+            {
+                _fields = Arrays.copyOf(_fields, _fields.length + 1);
+                System.arraycopy(_fields, _cursor, _fields, _cursor + 1, _size++);
```

Comment:
Should not the last parameter be `_size++ - _cursor`?
And yes, I would rather separate the `++` and `--` to different lines so that they are more comprehensible.
For example, it takes too much brain power to know if `5++ - 3` is `2` or `3`.

We have tests that call `add(HttpField)` but the error in the copy does not show because the array is initialized to 16 elements but in the tests _size is only 4-5.

Reply:
Yep it's a bug and if I change the capacity to 5 in the test I get a failure... but I don't see that it is the bug we are looking for :(